src/main/java/com/gregtechceu/gtceu/api/mui/widget/AbstractParentWidget.java

```diff
@@ -0,0 +1,117 @@
+package com.gregtechceu.gtceu.api.mui.widget;
+
+import com.gregtechceu.gtceu.api.mui.base.drawable.IDrawable;
+import com.gregtechceu.gtceu.api.mui.base.widget.IWidget;
+import com.gregtechceu.gtceu.api.mui.theme.WidgetTheme;
+import com.gregtechceu.gtceu.api.mui.widgets.VoidWidget;
+import com.gregtechceu.gtceu.client.mui.screen.ModularPanel;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.UnmodifiableView;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A widget which can hold any amount of children.
+ *
+ * @param <I> type of children (in most cases just {@link IWidget}). Use {@link VoidWidget} if no children should be
+ *            added.
+ * @param <W> type of this widget
+ */
+public class AbstractParentWidget<I extends IWidget, W extends AbstractParentWidget<I, W>> extends Widget<W> {
+
+    private final List<I> children = new ArrayList<>();
+
+    /**
+     * A list of all children of this widget. The list is modifiable contrary to the annotation.
+     * This just means that you shouldn't carelessly modify the list. Adding to the list also requires initialising the new child.
```

Comment:
Why not just make it unmodifiable then, and force users to use specific add/remove child functions?